warehouse/oidc/services.py

```diff
@@ -219,6 +220,28 @@ def _get_key_for_token(self, token):
         unverified_header = jwt.get_unverified_header(token)
         return self._get_key(unverified_header["kid"])
 
+    def token_identifier_exists(self, jti: str) -> bool:
+        """
+        Check if a JWT Token Identifier has already been used.
+        """
+        with redis.StrictRedis.from_url(self.cache_url) as r:
+            return bool(r.exists(f"/warehouse/oidc/{self.publisher}/{jti}"))
+
+    def store_jwt_identifier(self, jti: str, expiration: int) -> None:
+        """
+        Store the JTI with its expiration date if the key does not exist.
+        """
+        with redis.StrictRedis.from_url(self.cache_url) as r:
+            # Defensive: to prevent races, we expire the JTI slightly after
+            # the token expiration date. Thus, the lock will not be
+            # released before the token invalidation.
+            r.set(
+                f"/warehouse/oidc/{self.publisher}/{jti}",
+                exat=expiration + 5,
+                value="placeholder",
```

Comment:
Nitpick: sentinel value can be an empty string or int instead to lower the memory cost per token here:

```suggestion
                value="",
```

or:

```suggestion
                value=1,
```